### PR TITLE
Added support for internal_access for C#

### PR DESCRIPTION
**Pull request by gvaish**
@jskeet

### CODE REVIEW

**Comment by haberman** [comment]
ok to test

File: src/google/protobuf/compiler/csharp/csharp_generator.cc [hunk 1]
```diff
@@ -81,6 +81,13 @@ bool Generator::Generate(
     } else if (options[i].first == "base_namespace") {
       cli_options.base_namespace = options[i].second;
       cli_options.generate_directories = true;
+    } else if(options[i].first == "internal_access") {
+      std::string value = options[i].second;
+      if(value == "false" || value == "no" || value == "") {
```
Comment (jskeet) [inline src/google/protobuf/compiler/csharp/csharp_generator.cc]: Looking at options in other languages, most boolean flags don't seem to support false/no. I'd *at least* suggest that "" should mean "true", but I'd be tempted to just have:

```
else if (options[i].first == "internal_access") {
  cli_options.internal_access = true;
}
```

(Note the space after the `if` by the way, for consistency of spacing.)

File: src/google/protobuf/compiler/csharp/csharp_source_generator_base.cc [hunk 2]
```diff
@@ -59,7 +59,10 @@ void SourceGeneratorBase::WriteGeneratedCodeAttributes(io::Printer* printer) {
 }
 
 std::string SourceGeneratorBase::class_access_level() {
-  return IsDescriptorProto(descriptor_) ? "internal" : "public";  // public_classes is always on.
+  if(IsDescriptorProto(descriptor_)) {
+    return "internal";
+  }
+  return this->options()->internal_access ? "internal": "public";
```
Comment (jskeet) [inline src/google/protobuf/compiler/csharp/csharp_source_generator_base.cc]: We could either remove the `IsDescriptorProto(descriptor_)` call here and just change the generate_protos.sh file, or we could use

```
return IsDescriptorProto(descriptor_) || this->options()->internal_access ? "internal" : "public";
```

I don't mind either, particularly - but I think it's clearer to have one joint condition than an early return for descriptor protos and then a conditional.

**Comment by jskeet** [comment]
Have now merged #1390 - if you could rebase to that (and adjust the layout of the new option accordingly) that would be great.

File: src/google/protobuf/compiler/csharp/csharp_generator.cc (hunk 1 above)
Comment (gvaish) [inline src/google/protobuf/compiler/csharp/csharp_generator.cc]: Makes sense (for parity). And also yes for ` (`

(I personally love true/false & yes/no, but that's me.)

File: src/google/protobuf/compiler/csharp/csharp_source_generator_base.cc (hunk 2 above)
Comment (gvaish) [inline src/google/protobuf/compiler/csharp/csharp_source_generator_base.cc]: The only reason to split into `if` and `ternary` was _long line`. I'll also prefer single line.

On that note - that's the standard for line wrap at in this project - 80/100/120/none?

Comment (jskeet) [inline src/google/protobuf/compiler/csharp/csharp_source_generator_base.cc]: 80, where possible. In terms of splitting, I'd probably do:

```
return IsDescriptorProto(descriptor_) || this->options()->internal_access
    ? "internal" : "public";
```

**Comment by gvaish** [comment]
Rebased.

File: src/google/protobuf/compiler/csharp/csharp_source_generator_base.cc [hunk 3]
```diff
@@ -60,7 +60,7 @@ void SourceGeneratorBase::WriteGeneratedCodeAttributes(io::Printer* printer) {
 }
 
 std::string SourceGeneratorBase::class_access_level() {
-  return IsDescriptorProto(descriptor_) ? "internal" : "public";  // public_classes is always on.
+  return (IsDescriptorProto(descriptor_) || this->options()->internal_access) ? "internal": "public";
```
Comment (jskeet) [inline src/google/protobuf/compiler/csharp/csharp_source_generator_base.cc]: Tiny nit if you have the energy: add a space before the colon.
(If not, I'll fix it up as part of another PR.)

Comment (gvaish) [inline src/google/protobuf/compiler/csharp/csharp_source_generator_base.cc]: Damn. That's by bad. It should have had colon.

**Comment by jskeet** [comment]
Looks good to me - I assume you've tried it on a set of protos and it does what you want (and they still build)?

Happy to merge when it goes green, possibly with the specified nit fixed.

**Comment by jskeet** [comment]
Aha - thanks for the nit fix :) Will merge when green.

**Comment by gvaish** [comment]
I tried with `address.proto` in examples and it worked fine.

```
DYLD_LIBRARY_PATH=../src/.libs/ ../src/.libs/protoc --csharp_out=../.gen/cs/ --csharp_opt= addressbook.proto
```

Output:

```
  /// <summary>Holder for reflection information generated from addressbook.proto</summary>
  [global::System.Diagnostics.DebuggerNonUserCodeAttribute()]
  public static partial class AddressbookReflection {
```

and

```
  [global::System.Diagnostics.DebuggerNonUserCodeAttribute()]
  public sealed partial class Person : pb::IMessage<Person> {
    private static readonly pb::MessageParser<Person> _parser = new pb::MessageParser<Person>(() => new Person());
```

while when using

```
DYLD_LIBRARY_PATH=../src/.libs/ ../src/.libs/protoc --csharp_out=../.gen/cs/ --csharp_opt=internal_access= addressbook.proto
```

Output:

```
  /// <summary>Holder for reflection information generated from addressbook.proto</summary>
  [global::System.Diagnostics.DebuggerNonUserCodeAttribute()]
  internal static partial class AddressbookReflection {
```

and

```
  [global::System.Diagnostics.DebuggerNonUserCodeAttribute()]
  internal sealed partial class Person : pb::IMessage<Person> {
    private static readonly pb::MessageParser<Person> _parser = new pb::MessageParser<Person>(() => new Person());
```

**Comment by jskeet** [comment]
Heads-up: I think this has broken generation somehow. Still not quite sure why, but it's generating *all* classes as internal for me right now. Will investigate and send a new PR...

**Pull request merged**